src/ProductConstructionService/ProductConstructionService.DependencyFlow/PullRequestBuilder.cs

```diff
@@ -265,15 +295,108 @@ private String GenerateCodeFlowDescriptionForSubscription(
+                stringBuilder.AppendLine($"- **{depUpdate.DependencyName}**: [{depUpdate.ToVersion}]({diffLink})");
+            }
+        }
+
+        if (dependencyCategories.RemovedDependencies.Count > 0)
+        {
+            stringBuilder.AppendLine();
+            stringBuilder.AppendLine("**Removed Dependencies**");
+            foreach (DependencyUpdateSummary depUpdate in dependencyCategories.RemovedDependencies)
+            {
+                stringBuilder.AppendLine($"- **{depUpdate.DependencyName}**: {depUpdate.FromVersion}");
+            }
+        }
+
+        if (dependencyCategories.UpdatedDependencies.Count > 0)
+        {
+            stringBuilder.AppendLine();
+            stringBuilder.AppendLine("**Updated Dependencies**");
+            foreach (DependencyUpdateSummary depUpdate in dependencyCategories.UpdatedDependencies)
+            {
+                string? diffLink = GetLinkForDependencyItem(repoUri, depUpdate);
+                stringBuilder.AppendLine($"- **{depUpdate.DependencyName}**: [from {depUpdate.FromVersion} to {depUpdate.ToVersion}]({diffLink})");
+            }
+        }
+        return stringBuilder.ToString();
+    }
+
+    private static DependencyCategories CreateDependencyCategories(List<DependencyUpdateSummary> dependencyUpdateSummaries)
+    {
+        List<DependencyUpdateSummary> newDependencies = new();
+        List<DependencyUpdateSummary> removedDependencies = new();
+        List<DependencyUpdateSummary> updatedDependencies = new();
+
+        foreach (DependencyUpdateSummary depUpdate in dependencyUpdateSummaries)
+        {
+            if (string.IsNullOrEmpty(depUpdate.FromVersion))
+            {
+                newDependencies.Add(depUpdate);
+            }
+            else if (string.IsNullOrEmpty(depUpdate.ToVersion))
+            {
+                removedDependencies.Add(depUpdate);
+            }
+            else
+            {
+                updatedDependencies.Add(depUpdate);
+            }
+        }
+
+        return new DependencyCategories(newDependencies, removedDependencies, updatedDependencies);
+    }
+
+    private static string? GetLinkForDependencyItem(string repoUri, DependencyUpdateSummary dependencyUpdateSummary)
+    {
+        if (!string.IsNullOrEmpty(dependencyUpdateSummary.FromCommitSha) &&
+            !string.IsNullOrEmpty(dependencyUpdateSummary.ToCommitSha))
+        {
+            return GetChangesURI(repoUri, dependencyUpdateSummary.FromCommitSha, dependencyUpdateSummary.ToCommitSha);
+        }
+
+        if (!string.IsNullOrEmpty(dependencyUpdateSummary.ToCommitSha))
+        {
+            if (repoUri.Contains("github.com"))
+            {
+                return $"{repoUri}/commit/{dependencyUpdateSummary.ToCommitSha}";
+            }
+            if (repoUri.Contains("dev.azure.com"))
+            {
+                return $"{repoUri}?_a=history&version=GC{dependencyUpdateSummary.ToCommitSha}";
+            }
+        }
```

Comment:
I think this could cover easily for all cases (removal and addition):

```suggestion
        var sha = dependencyUpdateSummary.FromCommitSha ?? dependencyUpdateSummary.ToCommitSha;
        if (repoUri.Contains("github.com"))
        {
            return $"{repoUri}/commit/{sha}";
        }
        if (repoUri.Contains("dev.azure.com"))
        {
            return $"{repoUri}?_a=history&version=GC{sha}";
        }
```

Reply:
Here's what I want it to be known that the method does:
If dependencyUpdate has both From and To, return a link to a comparison between commits
If it has only To, return a link to that single commit
If not, return null, because we don't have a case for dependency updates where the "To" is null

I want the method to make those 3 cases as visible as possible. Since you don't like the nesting I added an extra method to distinguish between github and azdo url's (PS - if this becomes bothersome enough one day, we could create some kind of class like RepoURL, that holds the url string, and has methods to return useful links like commit links and comparison links, so that the logic could all be kept in one place)

Reply:
This was not about me not liking the nesting but rather having From only case. But if that's not useful then let's not have it.